docs/Storage Rules.md

```diff
@@ -0,0 +1,190 @@
+# Storage Rules
+
+Storage rules are based on Firebase. You can see the [pull request here](https://github.com/nhost/hasura-backend-plus/pull/144).
+
+The rules are governed by a `yaml` file, which defines routes and permissions, with authenticating functions. You can see [the template file here](https://github.com/nhost/hasura-backend-plus/blob/master/custom/storage-rules/rules.yaml)
+
+``` yaml
+functions:
+  isAuthenticated: 'return !!request.auth'
+  isOwner: "return !!request.auth && userId === request.auth['user-id']"
+  validToken: 'return request.query.token === resource.Metadata.token'
+paths:
+  /user/:userId/:
+    list: 'isOwner(userId)'
+  /user/:userId/:fileId:
+    read: 'isOwner(userId) || validToken()'
+    write: 'isOwner(userId)'
+```
+
+In the permission variables cookie returned by the [login]() or [refresh]() endpoints.
+
+This is a URL-encoded string, in the following template:
+
+``` txt
+s:<request.auth>.<checksum>
+```
+
+The request.auth part is a JSON object, in the following layout:
+
+``` json
+{
+  "user-id": "73f5d02c-484a-4003-98e4-bad5c6001882",
+  "allowed-roles": [
+    "user"
+  ],
+  "default-role": "user"
+}
+```
+
+You can use the `request.auth` parts in your storage rule functions.
+
+## Paths
+
+Paths allow you to define where your files are, and what permissions should apply to them. You can have dynamic paths and static paths.
+
+Dynamic paths can be used as variables within functions.
+
+### Folder paths
+
+Folder paths
+
+``` yaml
+paths:
+  /user/:userId/files/:
+    # Rules
+```
+
+**Note the trailing slash**. This is how you define folder paths.
+
+You can use this to add permissions to listing files in a directory or downloading a `.zip` file of all the files.
+
+### File paths
+
+File paths define rules for the individual files within your storage.
+
+``` yaml
+paths:
+  /user/:userId/files/:fileId:
+    # Rules
+```
+
+Here, the `/user` and `/files` parts are static paths. The `:userId` and `:fileId` parts are dynamic paths.
+
+Here's an example path that would be validated by this rule:
+
+``` txt
+/user/1/files/image.png
+```
+
+## Rules
+
+You can specify the following rules in your `rules.yaml` file.
+
+- create: upload a file and set metadata
+- update: update the file and patch metadata
+- get: return the file
+- list: return a zip file with all the permitted files with a key starting with /custom-path/
+- delete: delete the file
+
+For simple allow/deny, you can return boolean values. This should be a simple javascript expression in a string.
+
+For any complex permissions using variables, you should use [functions](#functions)
+
+``` yaml
+paths:
+  /public:
+    list: 'return true'
+  /private:
+    list: 'return false'
+```
+
+### File tokens
+
+Files are uploaded with a token within their metadata. This will mean that files are visible without authentication, if the file token is passed in as a query parameter.
```

Comment:
When a files is being uploaded HBP will automatically add a token to the files metadata. This token are ment as an access token and can be used with a `validToken`-function for the file to be accessable to users who are not authenticated. You can think of it as "if someone knows the token, they are allowed to read the file".

(Just example text, you are free to edit as you wish).

Reply:
I have updated this, is it clearer?

Reply:
Yes, looks good!